src/zjs_net.c

```diff
@@ -0,0 +1,1111 @@
+
+    if (handle) {
+        start_socket_timeout(handle, handle->timeout);
+
+        uint32_t len = net_nbuf_appdatalen(buf);
+        uint8_t *data = net_nbuf_appdata(buf);
+
+        DBG_PRINT("received data, context=%p, data=%p, len=%u\n", context, data, len);
+        // TODO: Probably not the right way to check if a socket is closed
+        if (len == 0 && data == NULL) {
+            // socket close
+            DBG_PRINT("closing socket, context=%p, socket=%u\n", context, handle->socket);
+            zjs_trigger_event(handle->socket, "close", NULL, 0, post_closed, handle);
+            return;
+        }
+
+        if (len && data) {
+            memcpy(handle->wptr, data, len);
+            handle->wptr += len;
+
+            // if not paused, call the callback to get JS the data
+            if (!handle->paused) {
+                handle->tcp_read_id = zjs_add_c_callback(handle, tcp_c_callback);
+                zjs_signal_callback(handle->tcp_read_id, NULL, 0);
+
+                DBG_PRINT("data recieved on context %p: data=%p, len=%u\n", context, data, len);
+            }
+        }
+
+        net_buf_unref(buf);
+    }
+}
+
+static inline void pkt_sent(struct net_context *context,
+                            int status,
+                            void *token,
+                            void *user_data)
+{
+    if (!status) {
+        int sent = POINTER_TO_UINT(token);
+        DBG_PRINT("Sent %d bytes\n", sent);
+        if (sent) {
+            zjs_callback_id id = POINTER_TO_INT(user_data);
+            if (id != -1) {
+                zjs_signal_callback(id, NULL, 0);
+            }
+        }
+    }
+}
+
+/**
+ * Write data to a socket
+ *
+ * @name write
+ * @memberof Net.Socket
+ * @param {Buffer} buf - Buffer being written to the socket
+ * @param {function=} func - Callback called when write has completed
+ */
+static jerry_value_t socket_write(const jerry_value_t function_obj,
+                                  const jerry_value_t this,
+                                  const jerry_value_t argv[],
+                                  const jerry_length_t argc)
+{
+    ZJS_VALIDATE_ARGS(Z_OBJECT, Z_OPTIONAL Z_FUNCTION);
+
+    sock_handle_t *handle = NULL;
+
+    jerry_get_object_native_handle(this, (uintptr_t *)&handle);
+    if (!handle) {
+        return zjs_error("socket handle not found");
+    }
+    start_socket_timeout(handle, handle->timeout);
+
+    zjs_buffer_t *buf = zjs_buffer_find(argv[0]);
+    struct net_buf *send_buf;
+    send_buf = net_nbuf_get_tx(handle->tcp_sock, K_NO_WAIT);
+    if (!send_buf) {
+        ERR_PRINT("cannot acquire send_buf\n");
+        return jerry_create_boolean(false);
+    }
+
+    bool status = net_nbuf_append(send_buf, buf->bufsize, buf->buffer, K_NO_WAIT);
+    if (!status) {
+        net_nbuf_unref(send_buf);
+        ERR_PRINT("cannot populate send_buf\n");
+        return jerry_create_boolean(false);
+    }
+
+    zjs_callback_id id = -1;
+    if (argc > 1) {
```

Comment:
again using an optional you should really refer to optcount to be sure it was a match w/ function